pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,284 @@
+
+        phi_real = qml.math.reshape(qml.math.real(phi), (dim,))
+        phi_imag = qml.math.reshape(qml.math.imag(phi), (dim,))
+        diag_value = prefactor_1 ** 2 * (
+            qml.math.dot(phi_real, phi_real) + qml.math.dot(phi_imag, phi_imag)
+        )
+        L = qml.math.scatter_element_add(L, (j, j), diag_value)
+
+        lam = psi * 1.0
+        lam_real = qml.math.reshape(qml.math.real(lam), (dim,))
+        lam_imag = qml.math.reshape(qml.math.imag(lam), (dim,))
+        # this entry is missing a factor of 1j
+        value = prefactor_1 * (qml.math.dot(lam_real, phi_real) + qml.math.dot(lam_imag, phi_imag))
+        T = qml.math.scatter_element_add(T, (j,), value)
+
+        for i in range(j - 1, -1, -1):
+            # after first iteration of inner loop: apply U_{i+1}^\dagger
+            if i < j - 1:
+                trainable_operations[i + 1].inv()
+                phi = device._apply_operation(phi, trainable_operations[i + 1])
+                trainable_operations[i + 1].inv()
+            # apply V_{i}^\dagger
+            phi = _apply_any_operation(phi, group_after_trainable_op[i], device, invert=True)
+            lam = _apply_any_operation(lam, group_after_trainable_op[i], device, invert=True)
+            inner_op = trainable_operations[i]
+            # extract and apply G_i
+            generator_2, prefactor_2 = _get_generator(inner_op)
+            # this state vector is missing a factor of 1j * prefactor_2
+            mu = device._apply_unitary(lam, qml.math.convert_like(generator_2, lam), inner_op.wires)
+            phi_real = qml.math.reshape(qml.math.real(phi), (dim,))
+            phi_imag = qml.math.reshape(qml.math.imag(phi), (dim,))
+            mu_real = qml.math.reshape(qml.math.real(mu), (dim,))
+            mu_imag = qml.math.reshape(qml.math.imag(mu), (dim,))
+            # this entry is missing a factor of 1j * (-1j) = 1, i.e. none
+            value = (
+                prefactor_1
+                * prefactor_2
+                * (qml.math.dot(mu_real, phi_real) + qml.math.dot(mu_imag, phi_imag))
+            )
+            L = qml.math.scatter_element_add(
+                L, [(i, j), (j, i)], value * qml.math.convert_like(qml.math.ones((2,)), value)
+            )
+            # apply U_i^\dagger
+            inner_op.inv()
+            lam = device._apply_operation(lam, inner_op)
+            inner_op.inv()
+
+        # apply U_j
+        psi = device._apply_operation(psi, outer_op)
+        # apply V_j
+        psi = _apply_any_operation(psi, group_after_trainable_op[j], device)
+
+    # postprocessing: combine L and T into the metric tensor.
+    # We require outer(conj(T), T) here, but as we skipped the factor 1j above,
+    # the stored T is real-valued. Thus we have -1j*1j*outer(T, T) = outer(T, T)
+    metric_tensor = L - qml.math.tensordot(T, T, 0)
+
+    return metric_tensor
+
+
+def _adjoint_metric_tensor_qnode(qnode, device, hybrid):
```

Comment:
I wonder, can you re-use the wrapper from the other `metric_tensor` function?

Reply:
I'll check this option. I thought that no, but forgot for what reason. :grimacing:

Reply:
Actually, there are quite some things going on in the `metric_tensor` QNode wrapper that are not applicable here. In particular, the tape-level function is _not_ a transform but a function that directly outputs the metric tensor!

Reply:
One could consider writing a helper function that only takes care of the actual contraction of the metric tensor with the classical Jacobian? That would separate the structural code from the math part kind of nicely :)

Reply:
Good idea 🙂